File: LC/[2411] Smallest Subarrays With Maximum Bitwise OR.cpp
class Solution {
public:
    vector<int> smallestSubarrays(vector<int>& nums)
    {
        int n = nums.size();
        int j = n-1;
        vector<int>rets(n);
        vector<int>count(32); //32 bit array indicate the current OR condition

        for (int i=n-1; i>=0; i--) //for each element in nums
        {
            for (int k=0; k<32; k++) //accumulate the bits into count
                count[k] += ((nums[i]>>k)&1);


            while (j>i && isOK(nums[j], count)) //can we take out num[jj] bits from counts and all positive bits still positive?  if so, we can remove nums[jj]
            {
                for (int k=0; k<32; k++)
                    count[k] -= ((nums[j]>>k)&1);
                j--;
            }

            rets[i] = j-i+1;
        }
        return rets;

    }

    bool isOK(int num, vector<int>&count)
    {
        for (int k=0; k<32; k++)
        {
            if (count[k] > 0 && (count[k] - ((num>>k)&1) <= 0))
                return false;
        }
        return true;
    }
};
Footer
© 2022 GitHub, Inc.
Footer navigation
Terms

File: LC/[2411] Smallest Subarrays With Maximum Bitwise OR.py
class Solution:
    def smallestSubarrays(self, nums: List[int]) -> List[int]:
        """

        """
        bit_cnt=[0]*32

        N=len(nums)
        res=[0]*N

        jj=N-1

        def is_ok_to_remove(nn, bit_cnt):
            for ii in range(32):
                if bit_cnt[ii] > 0 and bit_cnt[ii] - (nn>>ii & 1) <= 0: return False
            return True


        for ii in reversed(range(N)):
            #accumulate bits from (N-1)~ii
            for kk in range(32):
                bit_cnt[kk] += ((nums[ii] >> kk) & 1)

            while jj > ii and is_ok_to_remove(nums[jj], bit_cnt):
                for kk in range(32):
                    bit_cnt[kk] -= ((nums[jj]>>kk) & 1)
                jj -= 1

            res[ii]=jj-ii+1

        return res
